Astrometry db v1

stsci-hack opened the pull request:
Initial PR to implement new functionality in stwcs to apply headerlets from an astrometry database available through a web-service from MAST.  This implementation includes adding this functionality to the updatewcs function called by the HST calibration pipeline.

This should address all the requirements specified in issue #38 .

nden commented:
@stsci-hack 
Can you come up with a test for this change?
Also it appears `requests` now needs to be added as a test and install dependency.

nden commented:
`lxml` needs to be added as a dependence as well.

nden commented on stwcs/updatewcs/astrometry_utils.py:
```diff
@@ -0,0 +1,255 @@
+        for h in headerlets:
+            # Add solution as an alternate WCS
+            try:
+                headerlets[h].apply_as_alternate(fileobj, wcsname=h,
```
What is `wcsname` here? If  `h` is the headerlet, how does this work?

stsci-hack commented:
The latest commit corrects all the issues identified in the review.  This includes changing how the headerlet gets generated and how it gets appended to the science file (using attach_to_file instead of apply_as_alternate).  This new commit was used to successfully update an ACS/WFC exposure from the astrometry database.

stsci-hack commented:
The latest set of commits adds a new test module specific to the astrometry DB interface.  Another commit included revising updatewcs to make adding DB solutions options (through use of a new parameter) and therefore make requests and lxml optional dependencies (albeit turned on by default for pipeline use).  The exception handling was also revised to make the messages more human readable when they are thrown.

nden commented:
@stsci-hack Let me know when it's ready for a review.

stsci-hack commented:
@nden:  I just committed a final set of changes that add (I think) the dependencies, while also updating the tests to work with the logic for use in the pipeline.  With this latest commit, I believe I have fully addressed all the issues you raised.  Please go ahead and review this when you can.

stsci-hack commented:
@nden : Please merge this PR if you feel enough of the issues have been addressed to support initial testing by operations for the next build.  Any additional or remaining issues can be resolved upon my return from the holidays.  I realize a lot more documentation still needs to be added to explain this new step and how to control it, but that will have to make up the next PR.

nden commented on stwcs/updatewcs/astrometry_utils.py:
```diff
@@ -0,0 +1,436 @@
+            logger.warning('WARNING : AstrometryDB service inaccessible!')
+            logger.warning('    AstrometryDB called: {}'.format(
+                                self.serviceLocation))
+            logger.warning('    Status: {}'.format(r.status))
```
Since `r` is defined inside the `try` statement when an exception is raised it is not available, i.e. `r` is undefined at this point.

nden commented on stwcs/updatewcs/__init__.py:
```diff
@@ -93,6 +102,11 @@ def updatewcs(input, vacorr=True, tddcorr=True, npolcorr=True, d2imcorr=True,
+    if use_db:
+        # Establish any available connection to
+        #  an accessible astrometry web-service
+        astrometry = astrometry_utils.AstrometryDB()
```
`use_db` is true by default so calling `updatewcs` results always in an attempt to create a connection. You need to handle the case when a connection is not available so it doesn't crash. This is mostly why the tests are currently failing.

stsci-hack replied:
It turned out that there were problems with the logic that was used to interpret when to raise exceptions and when to perform the astrometry update at all.  Those errors were resolved and the tests for altwcs and updatewcs were updated to set the environment variable to turn off applying the astrometry update at all.  This now allows the tests to pass, and still provide the control needed by the pipeline for using this new code.  Please let me know if you feel there are any additional problems with this code that needs to be addressed.

nden commented:
@stsci-hack I am merging this so the nightly tests can run. 

I'm wondering though if the default value of `use_db` should be `False` instead of `True`. 
How is `updatewcs` supposed to be run by users now (not in the pipeline)?
If the intention is indeed to have this mode always on, then something should be added to the documentation.
